File: src/action/Action.cpp
#include "Action.h"

Action::Action(ActionType type) : is_valid_(true), type_(type) { }

std::vector<Action *> Action::parseActions(json msg)
{
	if (!msg.contains("actions"))
		return std::vector<Action *>();

	std::vector<Action *> actions;

	for (auto & actionJson : msg["actions"])
	{
		Action * newAction = nullptr;
		if (actionJson.contains("type"))
		{
			if (actionJson["type"] == "move")
				newAction = new MoveAction(actionJson);
			else if (actionJson["type"] == "create")
				newAction = new CreateAction(actionJson);
			else if (actionJson["type"] == "transfer_money")
				newAction = new TransferMoneyAction(actionJson);
			else if (actionJson["type"] == "build")
				newAction = new BuildAction(actionJson);
			if (newAction && !newAction->is_valid_)
			{
				delete newAction;
				newAction = nullptr;
			}
		}

		if (newAction != nullptr)
			actions.push_back(newAction);
	}

	return actions;
}

File: src/ts/replay_loader/action.ts
export interface BaseAction {
	type: string;
}
export interface MoveAction extends BaseAction {
	type: 'move';
	unit_id: number;
	x: number;
	y: number;
}
export interface TransferMoneyAction extends BaseAction {
	type: 'transfer_money';
	source_id: number;
	x: number;
	y: number;
	amount: number;
}
export interface BuildAction extends BaseAction {
	type: 'build';
	unit_id: number;
	x: number;
	y: number;
}
export interface CreateAction extends BaseAction {
	type: 'create';
	unit_type: number;
}
export interface AttackAction extends BaseAction {
	type: 'attack';
	unit_id: number;
	x: number;
	y: number;
	dmg: number;
}
export type TickAction = MoveAction | TransferMoneyAction | BuildAction | CreateAction | AttackAction;
